File: projects/7/BasicVMTranslator/internal/parser/parser.go
package parser

import (
	"fmt"
	"io"
	"ny/nand2tetris/basicvmtranslator/internal/config"
	"strconv"
	"strings"
)

type IParser interface {
	HasMoreLines() bool
	Advance() (bool, error)
	CommandType() config.CommandType
	Arg1() string
	Arg2() int
	RawLineLength() int
	CurrentLineIndex() int
}

type parser struct {
	lines            []string
	currentLineIndex int
	currentLine      string
	rawLineLength    int

	commandType config.CommandType
	arg1        string
	arg2        int
}

func NewParser(r io.Reader) (IParser, error) {
	content, err := io.ReadAll(r)
	if err != nil {
		return nil, err
	}

	if len(content) == 0 {
		return &parser{
			lines:            []string{},
			currentLineIndex: -1,
			commandType:      config.C_UNKNOWN,
		}, nil
	}

	rawLines := strings.Split(strings.ReplaceAll(string(content), "\r\n", "\n"), "\n")

	var lines []string
	for _, line := range rawLines {
		if commentIndex := strings.Index(line, "//"); commentIndex != -1 {
			line = line[:commentIndex]
		}
		line = strings.TrimSpace(line)
		if line != "" {
			lines = append(lines, line)
		}
	}

	return &parser{
		lines:            lines,
		rawLineLength:    len(rawLines),
		commandType:      config.C_UNKNOWN,
		currentLineIndex: -1,
	}, nil
}

func (p *parser) HasMoreLines() bool {
	return p.currentLineIndex+1 < len(p.lines)
}

func (p *parser) Advance() (bool, error) {
	if !p.HasMoreLines() {
		return false, nil
	}
	p.resetInstructionParts()
	p.currentLineIndex++
	p.currentLine = p.lines[p.currentLineIndex]

	parts := strings.Fields(p.currentLine)
	command := parts[0]

	commandType, commandArgLength := config.GetCommandInfo(command)
	if commandType == config.C_UNKNOWN {
		return true, fmt.Errorf("unknown command: %s", command)
	}

	if len(parts)-1 != commandArgLength {
		return true, fmt.Errorf("invalid number of arguments for command %s: expected %d, got %d", command, commandArgLength, len(parts)-1)
	}

	p.commandType = commandType
	if commandArgLength >= 1 {
		p.arg1 = parts[1]
	} else {
		p.arg1 = command
	}

	if commandArgLength == 2 {
		val, err := strconv.Atoi(parts[2])
		if err != nil {
			return true, fmt.Errorf("invalid argument for %s: %w", command, err)
		}
		p.arg2 = val
	} else {
		p.arg2 = 0
	}
	return true, nil
}

func (p *parser) CommandType() config.CommandType {
	return p.commandType
}
func (p *parser) Arg1() string {
	return p.arg1
}
func (p *parser) Arg2() int {
	return p.arg2
}

func (p *parser) resetInstructionParts() {
	p.currentLine = ""
	p.commandType = config.C_UNKNOWN
	p.arg1 = ""
	p.arg2 = 0
}

func (p *parser) RawLineLength() int {
	return p.rawLineLength
}

func (p *parser) CurrentLineIndex() int {
	return p.currentLineIndex
}

File: projects/8/FullVMTranslator/FullVMTranslator/Parser.cs
public interface IParser
{
    bool HasMoreLines();
    bool Advance();
    CommandType CommandType { get; }
    string Arg1 { get; }
    int Arg2 { get; }
    int RawLineLength();
    int CurrentLineIndex();
}

public class Parser : IParser
{
    private readonly List<string> lines;
    private int currentLineIndex;
    private string currentLine = "";
    private int rawLineLength;

    private CommandType commandType;
    private string arg1 = "";
    private int arg2;

    public Parser(TextReader reader)
    {
        var content = reader.ReadToEnd();
        if (string.IsNullOrEmpty(content))
        {
            lines = new List<string>();
            currentLineIndex = -1;
            commandType = CommandType.C_UNKNOWN;
            return;
        }

        var rawLines = content.Replace("\r\n", "\n").Split('\n');
        rawLineLength = rawLines.Length;

        lines = new List<string>();
        foreach (var line in rawLines)
        {
            var trimmed = line;
            var commentIndex = trimmed.IndexOf("//", StringComparison.Ordinal);
            if (commentIndex != -1)
            {
                trimmed = trimmed.Substring(0, commentIndex);
            }
            trimmed = trimmed.Trim();
            if (!string.IsNullOrEmpty(trimmed))
            {
                lines.Add(trimmed);
            }
        }
        commandType = CommandType.C_UNKNOWN;
        currentLineIndex = -1;
    }

    public bool HasMoreLines()
    {
        return currentLineIndex + 1 < lines.Count;
    }

    public bool Advance()
    {
        if (!HasMoreLines())
        {
            return false;
        }
        ResetInstructionParts();
        currentLineIndex++;
        currentLine = lines[currentLineIndex];

        var parts = currentLine.Split([' ', '\t'], StringSplitOptions.RemoveEmptyEntries);
        if (parts.Length == 0)
        {
            throw new Exception("empty command line");
        }

        var command = parts[0];

        var commandInfo = CommandConfig.GetCommandInfo(command);
        if (commandInfo.type == CommandType.C_UNKNOWN)
        {
            throw new Exception($"unknown command: {command}");
        }

        if (parts.Length - 1 != commandInfo.argLength)
        {
            throw new Exception($"invalid argument count for {command}: {currentLine}");
        }

        commandType = commandInfo.type;
        arg1 = commandInfo.argLength >= 1 ? parts[1] : command;
        arg2 = commandInfo.argLength == 2 ? int.Parse(parts[2]) : 0;
        return true;
    }

    public CommandType CommandType => commandType;
    public string Arg1 => arg1;
    public int Arg2 => arg2;

    private void ResetInstructionParts()
    {
        currentLine = "";
        commandType = CommandType.C_UNKNOWN;
        arg1 = "";
        arg2 = 0;
    }

    public int RawLineLength() => rawLineLength;
    public int CurrentLineIndex() => currentLineIndex;
}
